File: TwoPointer/1248. Count Number of Nice Subarrays/1248. Count Number of Nice Subarrays.java
class Solution {
    // slide window
    // two pointer
    public int numberOfSubarrays(int[] nums, int k) {
        List<Integer> list = new ArrayList<>();
        int n = nums.length;
        list.add(-1);
        for (int i = 0; i < n; i++) {
            if (nums[i]%2 == 1) list.add(i);
        }
        list.add(n);
        
        int rets = 0;
        if (list.size()-1<= k) return rets;

        for (int i = k+1; i < list.size(); i++) {
            rets += (list.get(i)-list.get(i-1))*(list.get(i-k)-list.get(i-k-1));
        }
        return rets;
    }
}

//  X X X X O X X X O X X X
//  ---------
//     disa
//                  --------
//                     disb
                    
// ret += disa * disb

File: TwoPointer/1248. Count Number of Nice Subarrays/1248. Count Number of Nice Subarrays.cpp
class Solution {
public:
    int numberOfSubarrays(vector<int>& nums, int k) {
        vector<int>oddidx;
        int n = nums.size();
        oddidx.push_back(-1);
        for (int i = 0; i < n; i++) {
            if (nums[i]%2 == 1) oddidx.push_back(i);
        }
        oddidx.push_back(n);
        
        int rets = 0;
        if (oddidx.size()-1<= k) return rets;

        for (int i = k+1; i < oddidx.size(); i++) {
            rets += (oddidx[i]-oddidx[i-1])*(oddidx[i-k]-oddidx[i-k-1]);
        }
        return rets;
    }
};
